File: native/windows/TlsSetupProcess.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace WorkspaceObservatory;

internal sealed record TlsSetupReply(int Id, string Status, string? Invitation, string? PeerCertificateSha256)
{
    internal static TlsSetupReply Parse(string raw)
    {
        if (Encoding.UTF8.GetByteCount(raw) > 8192 || JsonNode.Parse(raw) is not JsonObject value ||
            value.Any(field => field.Key is not ("id" or "status" or "invitation" or "peerCertificateSha256")) ||
            value["id"] is not JsonValue idValue || !idValue.TryGetValue<int>(out var id) || id < 1 ||
            value["status"]?.GetValue<string>() is not string status ||
            status is not ("idle" or "working" or "waiting" or "confirming" or "inactive" or "hosting" or "cancelled" or
                "configuration-ready" or "awaiting-confirmation" or "local-ready" or "acknowledged" or "unavailable"))
            throw new InvalidOperationException("Invalid setup reply.");
        var invitation = value["invitation"]?.GetValue<string>();
        var fingerprint = value["peerCertificateSha256"]?.GetValue<string>();
        if ((status == "hosting") != (invitation is not null) ||
            invitation is not null && (Encoding.UTF8.GetByteCount(invitation) > 2048 || !invitation.StartsWith("observatory-pair:v1:", StringComparison.Ordinal)) ||
            fingerprint is not null && !Regex.IsMatch(fingerprint, "\\A[a-f0-9]{64}\\z"))
            throw new InvalidOperationException("Invalid setup reply.");
        return new(id, status, invitation, fingerprint);
    }

    internal static void SelfTest()
    {
        if (Parse("{\"id\":1,\"status\":\"idle\",\"peerCertificateSha256\":null}").Status != "idle") throw new Exception("Setup reply failed.");
        foreach (var raw in new[] { "{}", "{\"id\":true,\"status\":\"idle\"}", "{\"id\":1,\"status\":\"paired\"}",
            "{\"id\":1,\"status\":\"hosting\"}", "{\"id\":1,\"status\":\"idle\",\"key\":\"private\"}" })
        {
            var rejected = false;
            try { Parse(raw); } catch { rejected = true; }
            if (!rejected) throw new Exception("Unsafe setup reply accepted.");
        }
        Console.WriteLine("Native setup reply validation passed.");
    }
}

// One native setup window owns one child. Private invitation text travels only
// through these pipes, never shell arguments, a webview or diagnostic output.
internal sealed class TlsSetupProcess : IAsyncDisposable
{
    private readonly Process process;
    private readonly ConcurrentDictionary<int, TaskCompletionSource<TlsSetupReply>> pending = new();
    private readonly SemaphoreSlim writer = new(1, 1);
    private readonly object gate = new();
    private readonly Task reader;
    private readonly Task errors;
    private Task? disposal;
    private int nextId, closed;
    internal bool ExitVerified { get; private set; }

    internal TlsSetupProcess(string runtime) : this(runtime,
        Path.Combine(AppContext.BaseDirectory, "Runtime", "node.exe"),
        Path.Combine(AppContext.BaseDirectory, "Collector", "scripts", "peer-tls-control.mjs")) { }

    internal TlsSetupProcess(string runtime, string node, string script)
    {
        if (!Path.IsPathFullyQualified(runtime) || !Directory.Exists(runtime) ||
            File.GetAttributes(runtime).HasFlag(FileAttributes.ReparsePoint) ||
            !Path.IsPathFullyQualified(node) || !File.Exists(node) || !Path.GetFileName(node).Equals("node.exe", StringComparison.OrdinalIgnoreCase) ||
            !Path.IsPathFullyQualified(script) || !File.Exists(script))
            throw new InvalidOperationException("Setup tools unavailable.");
        var start = new ProcessStartInfo(node) { UseShellExecute = false, CreateNoWindow = true,
            RedirectStandardInput = true, RedirectStandardOutput = true, RedirectStandardError = true,
            StandardInputEncoding = new UTF8Encoding(false, true), StandardOutputEncoding = new UTF8Encoding(false, true),
            WorkingDirectory = runtime };
        start.ArgumentList.Add(script); start.ArgumentList.Add("--runtime"); start.ArgumentList.Add(Path.GetFullPath(runtime));
        foreach (var key in new[] { "NODE_OPTIONS", "NODE_EXTRA_CA_CERTS", "NODE_PATH" }) start.Environment.Remove(key);
        process = new Process { StartInfo = start };
        if (!process.Start()) throw new InvalidOperationException("Setup process unavailable.");
        reader = ReadReplies();
        errors = DiscardErrors();
    }

    internal async Task<TlsSetupReply> Send(JsonObject command, CancellationToken cancellation = default)
    {
        int id;
        var completion = new TaskCompletionSource<TlsSetupReply>(TaskCreationOptions.RunContinuationsAsynchronously);
        lock (gate)
        {
            if (closed != 0 || pending.Count >= 8 || nextId >= 512) throw new InvalidOperationException("Setup process unavailable.");
            id = ++nextId; pending[id] = completion;
        }
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellation);
        timeout.CancelAfter(TimeSpan.FromSeconds(45));
        try
        {
            var line = new JsonObject { ["id"] = id, ["command"] = command.DeepClone() }.ToJsonString();
            if (Encoding.UTF8.GetByteCount(line) > 8192) throw new InvalidOperationException("Setup command too large.");
            await writer.WaitAsync(timeout.Token);
            try
            {
                await process.StandardInput.WriteLineAsync(line.AsMemory(), timeout.Token);
                await process.StandardInput.FlushAsync(timeout.Token);
            }
            finally { writer.Release(); }
            return await completion.Task.WaitAsync(timeout.Token);
        }
        catch { Stop(); throw new InvalidOperationException("Setup command unavailable."); }
        finally { pending.TryRemove(id, out _); }
    }

    private async Task ReadReplies()
    {
        var chars = new char[1024]; var frame = new StringBuilder();
        try
        {
            int count;
            while ((count = await process.StandardOutput.ReadAsync(chars)) != 0)
                for (var i = 0; i < count; i++)
                {
                    if (chars[i] == '\n')
                    {
                        var reply = TlsSetupReply.Parse(frame.ToString()); frame.Clear();
                        if (!pending.TryRemove(reply.Id, out var completion)) throw new InvalidOperationException();
                        completion.TrySetResult(reply);
                    }
                    else { frame.Append(chars[i]); if (frame.Length > 8192) throw new InvalidOperationException(); }
                }
        }
        catch { }
        finally { Stop(); }
    }

    private async Task DiscardErrors()
    {
        var buffer = new char[1024];
        try { while (await process.StandardError.ReadAsync(buffer) != 0) { } }
        catch { }
    }

    private void Stop()
    {
        lock (gate)
        {
            if (closed != 0) return; closed = 1;
            foreach (var item in pending) item.Value.TrySetException(new InvalidOperationException("Setup process stopped."));
            pending.Clear();
        }
        try { if (!process.HasExited) process.Kill(entireProcessTree: true); } catch { }
    }

    public ValueTask DisposeAsync()
    {
        lock (gate) return new ValueTask(disposal ??= DisposeCore());
    }

    private async Task DisposeCore()
    {
        Stop();
        using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(5));
        await process.WaitForExitAsync(timeout.Token);
        await Task.WhenAll(reader, errors).WaitAsync(timeout.Token);
        ExitVerified = process.HasExited;
        process.Dispose();
    }

    internal static async Task BridgeSelfTest(string node)
    {
        var runtime = Path.Combine(Path.GetTempPath(), "observatory-native-control-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(runtime);
        try
        {
            var bridge = new TlsSetupProcess(runtime, node, Path.Combine(AppContext.BaseDirectory, "Collector", "scripts", "peer-tls-control.mjs"));
            await using (bridge)
            {
                if ((await bridge.Send(new JsonObject { ["action"] = "status" })).Status != "idle" ||
                    (await bridge.Send(new JsonObject { ["action"] = "cancel" })).Status != "cancelled")
                    throw new InvalidOperationException("Setup bridge status failed.");
                var replies = await Task.WhenAll(Enumerable.Range(0, 4).Select(_ => bridge.Send(new JsonObject { ["action"] = "status" })));
                if (replies.Any(reply => reply.Status != "idle") || replies.Select(reply => reply.Id).Distinct().Count() != 4)
                    throw new InvalidOperationException("Setup request correlation failed.");
            }
            await bridge.DisposeAsync();
            if (!bridge.ExitVerified) throw new InvalidOperationException("Setup bridge exit not verified.");
            Console.WriteLine("Windows setup process bridge passed.");
        }
        finally { Directory.Delete(runtime, recursive: false); }
    }
}

File: native/TLSSetupProcess.swift
import Foundation
import Darwin

struct TLSSetupReply: Decodable {
    let id: Int
    let status: String
    let invitation: String?
    let peerCertificateSha256: String?

    static func selfTest() {
        precondition((try? parse(Data("{\"id\":1,\"status\":\"idle\",\"peerCertificateSha256\":null}".utf8)))?.status == "idle")
        for invalid in ["{}", "{\"id\":true,\"status\":\"idle\"}", "{\"id\":1,\"status\":\"hosting\"}",
                        "{\"id\":1,\"status\":\"paired\"}", "{\"id\":1,\"status\":\"idle\",\"key\":\"private\"}"] {
            precondition((try? parse(Data(invalid.utf8))) == nil)
        }
    }

    static func parse(_ data: Data) throws -> TLSSetupReply {
        guard data.count <= 8192,
              let fields = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              Set(fields.keys).isSubset(of: ["id", "status", "invitation", "peerCertificateSha256"]) else {
            throw CocoaError(.fileReadCorruptFile)
        }
        let reply = try JSONDecoder().decode(TLSSetupReply.self, from: data)
        guard reply.id > 0, ["idle", "working", "waiting", "confirming", "inactive", "hosting", "cancelled",
            "configuration-ready", "awaiting-confirmation", "local-ready", "acknowledged", "unavailable"].contains(reply.status),
              (reply.status == "hosting") == (reply.invitation != nil) else { throw CocoaError(.fileReadCorruptFile) }
        if let invitation = reply.invitation {
            guard invitation.utf8.count <= 2048, invitation.hasPrefix("observatory-pair:v1:") else {
                throw CocoaError(.fileReadCorruptFile)
            }
        }
        if let fingerprint = reply.peerCertificateSha256 {
            guard fingerprint.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil else {
                throw CocoaError(.fileReadCorruptFile)
            }
        }
        return reply
    }
}

// Owned by one native setup window. Commands and invitations use private pipes,
// never shell arguments, webview messages, logs or a local HTTP endpoint.
final class TLSSetupProcess {
    typealias Completion = (Result<TLSSetupReply, Error>) -> Void
    private let process = Process(), input = Pipe(), output = Pipe()
    private let queue = DispatchQueue(label: "Observatory.TLSSetup")
    private let writer = DispatchQueue(label: "Observatory.TLSSetup.Write")
    private var pending: [Int: Completion] = [:]
    private var buffer = Data()
    private var nextID = 1
    private var closed = false
    var isRunning: Bool { process.isRunning }

    convenience init(runtime: URL, resources: URL) throws {
        try self.init(runtime: runtime, node: resources.appendingPathComponent("Runtime/node/bin/node"),
                      script: resources.appendingPathComponent("Collector/scripts/peer-tls-control.mjs"))
    }

    init(runtime: URL, node: URL, script: URL) throws {
        guard runtime.isFileURL, node.isFileURL, script.isFileURL,
              FileManager.default.isExecutableFile(atPath: node.path),
              FileManager.default.fileExists(atPath: script.path),
              let resolved = realpath(runtime.path, nil) else { throw CocoaError(.fileReadNoSuchFile) }
        defer { free(resolved) }
        process.executableURL = node
        process.arguments = [script.path, "--runtime", String(cString: resolved)]
        process.currentDirectoryURL = URL(fileURLWithPath: String(cString: resolved), isDirectory: true)
        var environment = ProcessInfo.processInfo.environment
        for key in ["NODE_OPTIONS", "NODE_EXTRA_CA_CERTS", "NODE_PATH"] { environment.removeValue(forKey: key) }
        process.environment = environment
        process.standardInput = input
        process.standardOutput = output
        process.standardError = FileHandle.nullDevice
        try process.run()
        output.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            self?.queue.async { [weak self] in self?.received(data) }
        }
        process.terminationHandler = { [weak self] _ in
            self?.queue.async { [weak self] in self?.stop() }
        }
    }

    func send(_ command: [String: Any], completion: @escaping Completion) {
        queue.async { [self] in
            guard !closed, process.isRunning, pending.count < 8, nextID <= 512 else {
                DispatchQueue.main.async { completion(.failure(CocoaError(.fileReadUnknown))) }; return
            }
            let id = nextID
            do {
                var data = try JSONSerialization.data(withJSONObject: ["id": id, "command": command])
                guard data.count <= 8192 else { throw CocoaError(.fileReadTooLarge) }
                data.append(10)
                nextID += 1
                pending[id] = completion
                queue.asyncAfter(deadline: .now() + 45) { [weak self] in
                    guard let self, self.pending[id] != nil else { return }
                    self.stop()
                }
                let bytes = data
                writer.async { [weak self] in
                    guard let self else { return }
                    do { try self.input.fileHandleForWriting.write(contentsOf: bytes) }
                    catch { self.queue.async { [weak self] in self?.stop() } }
                }
            } catch {
                if pending[id] != nil { stop() }
                else { DispatchQueue.main.async { completion(.failure(CocoaError(.fileWriteUnknown))) } }
            }
        }
    }

    func close() { queue.async { [weak self] in self?.stop() } }

    private func received(_ data: Data) {
        guard !closed else { return }
        guard !data.isEmpty, buffer.count + data.count <= 16384 else { stop(); return }
        buffer.append(data)
        while let end = buffer.firstIndex(of: 10) {
            let frame = Data(buffer[..<end])
            buffer.removeSubrange(...end)
            do {
                let reply = try TLSSetupReply.parse(frame)
                guard let completion = pending.removeValue(forKey: reply.id) else { stop(); return }
                DispatchQueue.main.async { completion(.success(reply)) }
            } catch { stop(); return }
        }
    }

    private func stop() {
        guard !closed else { return }
        closed = true
        if process.isRunning {
            process.terminate()
            let child = process
            DispatchQueue.global().asyncAfter(deadline: .now() + 2) {
                if child.isRunning { kill(child.processIdentifier, SIGKILL) }
            }
        }
        output.fileHandleForReading.readabilityHandler = nil
        try? input.fileHandleForWriting.close()
        try? output.fileHandleForReading.close()
        buffer.removeAll(keepingCapacity: false)
        let callbacks = Array(pending.values)
        pending.removeAll()
        for completion in callbacks { DispatchQueue.main.async { completion(.failure(CocoaError(.fileReadUnknown))) } }
    }

    deinit {
        output.fileHandleForReading.readabilityHandler = nil
        try? input.fileHandleForWriting.close()
        try? output.fileHandleForReading.close()
        if process.isRunning { process.terminate() }
    }
}
